sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -3187,7 +3189,37 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
       ctx: PropertyListContext): Map[String, String] = withOrigin(ctx) {
     val properties = ctx.property.asScala.map { property =>
       val key = visitPropertyKey(property.key)
-      val value = visitPropertyValue(property.value)
+      // If the expression provided for the option value is a literal, use the string
+      // representation of the contents of the literal value provided here. Otherwise, if this
+      // expression is constant but non-literal, evaluate the expression and use its value instead.
+      val value = if (property.value == null) {
+        null
+      } else {
+        val parsed = expression(property.value)
+        def error: Throwable = {
+          new ParseException(
+            errorClass = "INVALID_SQL_SYNTAX",
+            messageParameters = Map(
+              "inputString" ->
+                s"option or property key $key is invalid; only constant expressions are supported"),
+            ctx)
+        }
+        val plan = try {
+          val analyzer: Analyzer = ResolveDefaultColumns.DefaultColumnAnalyzer
+          val analyzed = analyzer.execute(Project(Seq(Alias(parsed, "col")()), OneRowRelation()))
+          analyzer.checkAnalysis(analyzed)
+          ConstantFolding(analyzed)
+        } catch {
+          case _: AnalysisException => throw error
+        }
+        val result: Expression = plan.collectFirst {
+          case Project(Seq(a: Alias), OneRowRelation()) => a.child
+        }.get
+        result match {
+          case expr if expr.foldable => expr.eval().toString
```

Comment:
How about to wrap `expr` by the `Cast` to string expression and invoke `eval`. `toString` and casting to string can  output different results if we care of them.

Reply:
This is done.

Reply:
Update: I found that using `Cast` does not work in all cases, such as timestamps since we do not currently support casting timestamps to strings. I switched this back to use `.toString` and added a comment explaining this. It seems to work well for different types of expressions.